### PR TITLE
Do not read predictions in memory, only after score

### PR DESCRIPTION
**Pull request by franchuterivera**
- Code cleanup for _read_np_fn

- Fix the precision type when reading files, in _read_np_fn

- Score the predictions without reading them. Only read the actual file when we are sure it is the best prediction.

### CODE REVIEW

**Comment by franchuterivera** [comment]
I do have one extra open, and is the need of the complex data structure with loaded key.

That is, I don't see this "loaded" key being used at all, in the ensemble structure.

File: autosklearn/ensemble_builder.py [hunk 1]
```diff
@@ -950,13 +929,20 @@ def _delete_excess_models(self):
...
+        if path.endswith("gz"):
+            open_method = gzip.open
+        elif path.endswith("npy"):
+            open_method = open
         else:
-            predictions = np.load(fp, allow_pickle=True)
-        return predictions
+            raise ValueError("Unknown filetype %s" % path)
+        with open_method(path, 'rb') as fp:
+            if self.precision == 16:
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:940]: Hey, I don't know whether the string-based version is still used anywhere. Could you please change the code to check for both strings and integers, but also issue a warning if the precision is a string?

File: autosklearn/ensemble_builder.py [hunk 2]
```diff
@@ -411,40 +411,36 @@ def read_ensemble_preds(self):
...
-                            os.path.getmtime(y_ens_fn),
-                        )
-
-                    self.read_preds[y_ens_fn]["ens_score"] = score
-                    self.read_preds[y_ens_fn][Y_ENSEMBLE] = y_ensemble
-                    self.read_preds[y_ens_fn]["mtime_ens"] = os.path.getmtime(
-                        y_ens_fn
+                y_ensemble = self._read_np_fn(y_ens_fn)
+                score = calculate_score(solution=self.y_true_ensemble,
+                                        # y_ensemble = y_true for ensemble set
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:416]: Could you please drop this comment? I don't remember what it means.

File: autosklearn/ensemble_builder.py [hunk 3]
```diff
@@ -526,14 +526,7 @@ def get_n_best_preds(self):
...
             if self.read_preds[k][Y_ENSEMBLE] is None:
-                if k.endswith("gz"):
-                    open_method = gzip.open
-                elif k.endswith("npy"):
-                    open_method = open
-                else:
-                    raise ValueError("Unknown filetype %s" % k)
-                with open_method(k, 'rb') as fp:
-                    self.read_preds[k][Y_ENSEMBLE] = self._read_np_fn(fp=fp)
+                self.read_preds[k][Y_ENSEMBLE] = self._read_np_fn(k)
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:529]: I'm wondering whether this loading logic should go after the part which contains '#consider performance_range_threshold'?

Comment (mfeurer) [reply autosklearn/ensemble_builder.py:529]: Or even better after dropping models/loaded predictions. I guess that would be maximally memory efficient.

**Pull request merged**